kolibri/core/analytics/middleware.py

```
@@ -0,0 +1,135 @@
+    command_pid = 0
+
+    def process_request(self, request):
+        """
+        Store the start time, memory and load when the request comes in.
+        """
+        if not self.disabled:
+            self.metrics = Metrics()
+
+    def shutdown(self):
+        """
+        Disable this middleware and clean all the static variables
+        """
+        MetricsMiddleware.disabled = True
+        MetricsMiddleware.command_pid = 0
+        delattr(self, 'metrics')
+        if os.path.exists(PROFILE_LOCK):
+            try:
+                os.remove(PROFILE_LOCK)
+            except OSError:
+                pass  # lock file was deleted by other process
+
+    def check_start_conditions(self):
+        """
+        Do the needed checks to enable the Middleware if possible
+        """
+        if MetricsMiddleware.disabled and conf.OPTIONS["Server"]["PROFILE"]:
+            if os.path.exists(PROFILE_LOCK):
+                try:
+                    with open(PROFILE_LOCK, 'r') as f:
+                        MetricsMiddleware.command_pid = int(f.readline())
+                        if SUPPORTED_OS:
+                            MetricsMiddleware.disabled = False
+                except (IOError, TypeError, ValueError):
+                    # Kolibri command PID file has been deleted or it's corrupted
+                    try:
+                        os.remove(PROFILE_LOCK)
+                    except OSError:
+                        pass  # lock file was deleted by other process
+
+    def process_response(self, request, response):
+        """
+        Calculate and output the page generation details
+        Log output consist on:
+        Datetime, request path, request duration, memory before, memory after requests is finished,
+        cpu load before, cpu load after the request is finished, max
+        Being `max` True or False to indicate if this is the slowest request since logging began.
+        """
+        self.check_start_conditions()
```

I am not sure how much of a performance hit this is going to cause, but it seems like we are touching the disk for every request (regardless of whether the profiling is turned off or not), in order to be able to check if the profiling command is currently being run.

As running the profiling command will require being able to invoke the Kolibri command from the command line anyway, could we instead require that the profiling command is run before the server is started, and then do a startup check on the middleware in the `__init__` method: https://docs.djangoproject.com/en/1.11/topics/http/middleware/#marking-middleware-as-unused ?

---

Oh, sorry - I overstated the issue, as you are checking the conf file value. We could perhaps just throw `MiddlewareNotUsed` during `__init__` if the conf value is not set?

---

I am not sure if I u nderstand what you mean: do you want to throw an error when kolibri starts if the conf value is  not set? The idea was to leave it not set by default. Only if you are going to profile you need to activate it.

---

Forget my previous comment. I got it, and implemented it. It was a very good idea :100: